xonsh/main.py

```diff
@@ -441,10 +441,13 @@ def func_sig_ttin_ttou(n, f):
     shell = builtins.__xonsh__.shell
     history = builtins.__xonsh__.history
     exit_code = 0
+
+    if shell and not env["XONSH_INTERACTIVE"]:
+        shell.ctx.update({"exit": sys.exit})
+
     try:
         if args.mode == XonshMode.interactive:
             # enter the shell
-            env["XONSH_INTERACTIVE"] = True
```

Comment:
I think this line should probably stay because it is possible to call `main_xonsh()` without calling `premain()`, namely in the tests.